src/dispatcher.rs

```diff
@@ -61,41 +64,50 @@ impl<C> CommandDispatcher<C> {
     ///
     /// Unicode characters are currently not supported. This may be fixed in the future.
     pub fn dispatch(&self, ctx: &mut C, command: &str) -> bool {
-        let parsed = Self::parse_into_arguments(command);
+        // let parsed = Self::parse_into_arguments(command);
 
         let mut current_node = self.root;
 
-        for argument in &parsed {
+        let mut input = Input::new(command);
+
+        while !input.empty() {
             // try to find a node satisfying the argument
             let node = &self.nodes[current_node.0];
-
+            
             // TODO: optimize linear search using a hash-array mapped trie
-            if let Some(next) = node.next.iter().find(|next| {
+            if let Some((next, next_input)) = node.next.iter().filter_map(|next| {
                 let kind = &self.nodes[next.0].kind;
+                let mut input = input.clone();
+
+                &input;
 
-                match kind {
-                    NodeKind::Parser(parser) => parser.satisfies(ctx, argument),
-                    NodeKind::Literal(lit) => lit == argument,
+                if match kind {
+                    NodeKind::Parser(parser) => parser.satisfies(ctx, &mut input),
+                    NodeKind::Literal(lit) => lit == input.head(" "),
                     NodeKind::Root => unreachable!("root NodeKind outside the root node?"),
+                } {
+                    Some((next, input))
+                } else {
+                    None
                 }
-            }) {
+            }).next() {
                 current_node = *next;
+                input = next_input;
             } else {
                 return false;
             }
         }
 
         if let Some(exec) = &self.nodes[current_node.0].exec {
-            exec(ctx, &parsed);
+            exec(ctx, command);
             true
         } else {
             false
         }
     }
 
-    fn parse_into_arguments(command: &str) -> SmallVec<[&str; 4]> {
-        // TODO: proper parser with support for strings in quotes
-        command.split(" ").collect()
+    pub fn command_meta(&self) -> impl Iterator<Item = &CommandMeta> {
```

Comment:
Should be called `commands()` or `command_metas()` or similar—need to indicate plurality